ta4j-core/src/main/java/org/ta4j/core/rules/TrailingATRStopLossRule.java

```diff
@@ -0,0 +1,103 @@
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+package org.ta4j.core.rules;
+
+import org.ta4j.core.BarSeries;
+import org.ta4j.core.TradingRecord;
+import org.ta4j.core.indicators.ATRIndicator;
+import org.ta4j.core.indicators.helpers.*;
+import org.ta4j.core.num.Num;
+
+/**
+ * A rule for a trailing stop loss using the Average True Range (ATR) as a
+ * basis.
+ * <p>
+ * The rule is satisfied when the price drops below the highest price since the
+ * position entry (for a long position) or rises above the lowest price since
+ * the position entry (for a short position) by a certain multiple of the ATR.
+ */
+public class TrailingATRStopLossRule extends AbstractRule {
+
+    private final TransformIndicator atr;
+    private final ClosePriceIndicator closePrice;
+    private final HighPriceIndicator highPrice;
+    private final LowPriceIndicator lowPrice;
+
+    /**
+     * Constructor.
+     *
+     * @param series         the bar series
+     * @param barCount       the number of bars to consider for the ATR
+     * @param atrCoefficient the coefficient for the ATR
+     */
+    public TrailingATRStopLossRule(BarSeries series, int barCount, double atrCoefficient) {
+        if (barCount < 1) {
+            throw new IllegalArgumentException("barCount must be positive");
+        }
+        if (atrCoefficient < 0) {
+            throw new IllegalArgumentException("atrCoefficient cannot be negative");
+        }
+
+        this.atr = TransformIndicator.multiply(new ATRIndicator(series, barCount), atrCoefficient);
+        this.closePrice = new ClosePriceIndicator(series);
+        this.highPrice = new HighPriceIndicator(series);
+        this.lowPrice = new LowPriceIndicator(series);
+    }
+
+    @Override
+    /**
+     * Checks if the rule is satisfied.
+     *
+     * @param index         the bar index
+     * @param tradingRecord the current trading record
+     * @return true if the rule is satisfied, false otherwise
+     */
+    public boolean isSatisfied(int index, TradingRecord tradingRecord) {
+        // Rule is not satisfied if there is no active trade
+        if (tradingRecord == null || tradingRecord.isClosed()) {
+            return false;
+        }
+
+        // Calculate the number of bars since the entry of the current trade
+        int barsSinceEntry = index - tradingRecord.getCurrentPosition().getEntry().getIndex() + 1;
+        Num currentPrice = this.closePrice.getValue(index);
+
+        // Determine if stop loss should be triggered for a long position
+        if (tradingRecord.getCurrentPosition().getEntry().isBuy()) {
+            HighestValueIndicator highestPrice = new HighestValueIndicator(this.highPrice, barsSinceEntry);
```

Comment:
@hhashim1 are you planning to use this trailing ATR stop loss rule in an offline back-testing context or live, real-time trading? The current implementation is designed for the latter as it tries to solve for the scenario where the stop loss threshold gets breached "in-between" bars. If this isn't needed then I can change this to a more conventional approach.

Reply:
> @hhashim1 are you planning to use this trailing ATR stop loss rule in an offline back-testing context or live, real-time trading? The current implementation is designed for the latter as it tries to solve for the scenario where the stop loss threshold gets breached "in-between" bars. If this isn't needed then I can change this to a more conventional approach.

@TheCookieLab The ATR Stop Loss is used in both backtesting and Live trading.